client/normal_value.go

```diff
@@ -0,0 +1,1531 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package client
```

Comment:
suggestion: This file is pretty hard to read, and I can see it getting very disorganized over time, and very hard to find stuff/see what exists.  I suggest splitting it up by type in its own directory within client:

``` client/value
void.go
nil.go
int.go
// etc
```
If you really want the types directly in the client package, you could always re-export them(or just the constructors+interface?) Rust style from a mod.rs style file (client/normal_value.go):
```
type NormalValue = value.NormalValue
func NewNormalInt(...) { // I don't know if there is a better way than this in Go for funcs?
  return value.NormalValue()
}
```

Reply:
I also thought about it but then it would need access to `client.Document` and `client` package will also depend on it. You have a suggestion?

Reply:
```
type NormalValue = value.NormalValue  
func NewNormalInt(...) { // I don't know if there is a better way than this in Go for funcs?  
  return value.NormalValue()  
} 
```

Please don't 😅.

I'm fine if we go for a client sub-package. `client/normalized` would be my choise. Then we could have the interface`normalized.Value` and the types `normalized.String`, `normalized.Int`, etc.

Reply:
> I also thought about it but then it would need access to client.Document and client package will also depend on it. You have a suggestion?

No need to bring `client.Document` into the sub package. Just have the `normilized.Value` interface implement for `client.Document` where `client.Document` is defined.

Reply:
Sorry. I need to point out that for the last comment to work we need to keep the interface in client. So the `normalized` package wouldn't have the `normalized.Value` interface. It would just hold the types.

Reply:
> No need to bring client.Document into the sub package. Just have the normilized.Value interface implement for client.Document where client.Document is defined.

Then it won't be possible to get normal value from `NewNormalValue` passing in a `*client.Document`. Not sure it's a good trade-off.

Reply:
An alternative could be to namespace the file names within the client package: `_normal_int.go`, `_normal_value.go`, etc - please dont do this without a namespace though, or we wont be able to find anything else in the directory.

Reply:
That might be the more sensible approach. No need for the `_` prefix though.

Reply:
is "normal" bad namespace? `normal_int.go`, `normal_nillable_int.go`...?

Reply:
> is "normal" bad namespace? `normal_int.go`, `normal_nillable_int.go`...?

I much prefer just `normal_...` 👍

Reply:
> I would prefer that they are all either at the end alphabetically, or at the start - otherwise the other files will be split on either side.

I'm pretty sure the compiler ignores files with `_` prefixes just like it ignores  files with a `.` prefix.

Reply:
I confirm https://github.com/golang/go/blob/master/src/go/build/build.go#L548

Reply:
:(

`zzz_normal_int.go` :laughing: ? I will have to tolerate them being in the middle of the package them :(